Tensor product distributes over direct product for finitely presented modules

maddycrim opened the pull request:
For `Module.FinitePresentation.exists_fin_exact` : If `M` is a finitely presented `R`-module then there exists an exact sequence `R^m → R^n → M → 0`.

For `tensorPi_equiv_piTensor'` : If `M` is a finitely presented `R`-module and `N i` is a family of `R`-modules then `M ⊗[R] (Π i, N i) ≃ₗ[R] Π i, (M ⊗[R] N i)`.

kbuzzard commented on FLT/DedekindDomain/FiniteAdeleRing/TensorPi.lean:
```diff
@@ -151,3 +192,105 @@ lemma tensorPi_equiv_piTensor_apply (m : M) (n : ∀ i, N i) :
+  set N2 := Π i, ((Fin n → R) ⊗[R] N i)
+  set N3 := Π i, (M ⊗[R] N i)
+  set N4 : Type u := PUnit
+  set N5 : Type u := PUnit
```
Clearly you have a plan here. Can you face drawing some ASCII art in a comment and explaining what is going on in this proof, so that people understand it? I know that mathlib doesn't do that sort of thing but in my mind that's a weakness of mathlib. Any proof which is beyond a certain length should in my opinion be explained a bit, and whatever that length is, this proof is definitely beyond it :-)

maddycrim replied:
Done! I removed all the `set` declarations for the modules as well. Since the diagram shows the actual modules instead of names like `M1`, `M2`, etc., I figured it was clearer to just use the full expressions directly. The original purpose of the `M1`, `M2`, etc. anyhow was to compare with commutative diagram from `Mathlib.Algebra.FiveLemma`.

kbuzzard reviewed (commented):
Thanks a lot! This is almost ready to go. I've made some hopefully helpful comments.

maddycrim commented on FLT/DedekindDomain/FiniteAdeleRing/TensorPi.lean:
```diff
@@ -22,7 +26,44 @@ If `M` is a finite free module and `Nᵢ` is an indexed collection of modules of
+
+  let g : (Fin n → R) →ₗ[R] M := iso.symm.toLinearMap.comp (Submodule.mkQ K)
+
+  have h₁ : LinearMap.range f = K := by
```
I've gone ahead and got rid of the double spacing. When solving the problem it helped me visually. I will keep this in mind for the future.

maddycrim commented on FLT/DedekindDomain/FiniteAdeleRing/TensorPi.lean:
```diff
@@ -22,7 +26,44 @@ If `M` is a finite free module and `Nᵢ` is an indexed collection of modules of
+    rw [← hS]
+    simp only [f, Fintype.range_linearCombination, gens, (Function.Surjective.range_comp
+      (Finset.equivFin S).symm.surjective Subtype.val).trans Subtype.range_val_subtype]
+    rfl
```
I got rid of the rw, changed the simp only to have \l hs, and adding Finset.setOf_mem to the simp only solved the goal so I could avoid using rfl.

maddycrim commented on FLT/DedekindDomain/FiniteAdeleRing/TensorPi.lean:
```diff
@@ -151,3 +192,105 @@ lemma tensorPi_equiv_piTensor_apply (m : M) (n : ∀ i, N i) :
+    refine ext' fun x y ↦ ?_
+    simp only [LinearMap.coe_comp, comp_apply, IsLinearMap.mk'_apply, i₂, i₁, g₁, N2, N1]
+    rw [LinearMap.rTensor_tmul]
+    erw [tensorPi_equiv_piTensor_apply, tensorPi_equiv_piTensor_apply]
```
Thank you so much for this information! I will keep this in mind going forward. I added LinearEquiv.coe_coe where necessary and was able to get rid of the erw. Interestingly, when I do simp? for hc1, LinearEquiv.coe_coe doesn't show, but adding it the simp only still fixed the issue.

maddycrim commented on FLT/DedekindDomain/FiniteAdeleRing/TensorPi.lean:
```diff
@@ -151,3 +192,105 @@ lemma tensorPi_equiv_piTensor_apply (m : M) (n : ∀ i, N i) :
+/-- Tensoring with a finitly presented module commutes with arbitrary products. -/
+noncomputable def tensorPi_equiv_piTensor' [Module.FinitePresentation R M] :
+    M ⊗[R] (Π i, N i) ≃ₗ[R] Π i, (M ⊗[R] N i) := by
+  choose n m f g exact surj using Module.FinitePresentation.exists_fin_exact R M
```
I went ahead and made the switch! I ideally would've used the obtain tactic to start, but this didn't work before I made the switch. Also, I tested it both ways, when going straight into tactic mode versus doing the term part in term mode then switching into tactic mode, the apply lemma was a simple rfl either way. This surprised me. When I originally did the free and finite case I had went straight into tactic mode which made it incredibly difficult to prove the apply lemma.

kbuzzard replied:
Yeah some tactics like `cases`  (which is what powers `obtain`) don't work on an "Exists" statement when the goal is a Type instead of a Prop. This is because `cases` internally runs induction on the inductive type `Exists` (defined in core Lean as
```
inductive Exists {α : Sort u} (p : α → Prop) : Prop where
  /-- Existential introduction. If `a : α` and `h : p a`,
  then `⟨a, h⟩` is a proof that `∃ x : α, p x`. -/
  | intro (w : α) (h : p w) : Exists p
```
) and when that definition is run, it generates a recursor `Exists.rec`, the principle of recursion for Exists. When you define the natural numbers you get `Nat.rec` automatically defined in the same way, and `Nat.rec` is just the usual principle of recursion for naturals, saying things like "a valid way to define a function F from the naturals to the reals is by saying "send 0 to 37 and n+1 to sqrt(F(n))". In particular Nat.rec allows you to create data. More precisely if you `#check Nat.rec` then you'll see that the "motive" variable has type `Nat -> Sort u`, here "Sort" means "either Type or Prop". However the "motive" for Exists.rec is `Exists p -> Prop` i.e. it does not allow Type, which means that when you use the `cases` tactic and it tries to unravel it into an actual way of deducing (old hypotheses -> goal) from (new hypotheses -> goal) it has to make the motive have type `Exists p -> type of goal` so if the type of the goal isn't Prop it fails.

That's what's going on internally. The reason for this restriction on `Exists.rec` is that if you let Exists eliminate into Type then you can prove contradictions! Basically you can prove "there exists an even natural" with 2, and then prove "there exists an even natural" again with 4, and one of the axioms of Lean is that these two proofs are equal, because all proofs of a theorem are equal in Lean (the proof is `rfl`), and if you could eliminate an existence into Type you could deduce that 2 = 4, which isn't good.

maddycrim commented on FLT/DedekindDomain/FiniteAdeleRing/TensorPi.lean:
```diff
@@ -151,3 +192,105 @@ lemma tensorPi_equiv_piTensor_apply (m : M) (n : ∀ i, N i) :
+    f₁ f₂ f₃ f₄ g₁ g₂ g₃ g₄ i₁ i₂ i₃ i₄ i₅
+    hc₁ hc₂ hc₃ hc₄ hf₁ hf₂ hf₃ hg₁ hg₂ hg₃ hi₁ hi₂ hi₄ hi₅
+
+  exact IsTensorProduct.equiv this
```
I proved `tensorPi_equiv_piTensor'_apply`.

maddycrim commented:
I can't remove the awaiting author label, but I've addressed the comments and pushed my changes.

> Thanks a lot! This is almost ready to go. I've made some hopefully helpful comments.

kbuzzard commented:
Yes sorry about that, the tag is used for my own internal algorithm for managing PRs, it basically means "the ball is not in my court at the minute, look at this PR again tomorrow to see if it's moved on".

kbuzzard commented on FLT/DedekindDomain/FiniteAdeleRing/TensorPi.lean:
```diff
@@ -22,7 +26,32 @@ If `M` is a finite free module and `Nᵢ` is an indexed collection of modules of
+    simp only [g, LinearMap.coe_comp, LinearEquiv.coe_coe, EquivLike.comp_surjective,
+      Submodule.mkQ_surjective]
+  exact ⟨n, m, f, g, exact_fg, this⟩
+
```
```suggestion
```
